circuit-shared-elements/src/commonMain/kotlin/com/slack/circuit/sharedelements/SharedElementTransitionScope.kt

```diff
@@ -0,0 +1,177 @@
+// Copyright (C) 2024 Slack Technologies, LLC
+// SPDX-License-Identifier: Apache-2.0
+package com.slack.circuit.sharedelements
+
+import androidx.compose.animation.AnimatedVisibilityScope
+import androidx.compose.animation.EnterExitState
+import androidx.compose.animation.ExperimentalSharedTransitionApi
+import androidx.compose.animation.SharedTransitionScope
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.CompositionLocalProvider
+import androidx.compose.runtime.ProvidableCompositionLocal
+import androidx.compose.runtime.compositionLocalOf
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateMapOf
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import com.slack.circuit.sharedelements.SharedElementTransitionScope.AnimatedScope
+import com.slack.circuit.sharedelements.SharedElementTransitionScope.AnimatedScope.Navigation
+import com.slack.circuit.sharedelements.SharedElementTransitionScope.AnimatedScope.Overlay
+
+/**
+ * [SharedElementTransitionScope] provides a [SharedTransitionScope] for the standard shared
+ * elements/shared bounds animations. This also provides access to a [AnimatedVisibilityScope],
+ * which can be retrieved using [getAnimatedScope] or [requireAnimatedScope].
+ *
+ * An [AnimatedScope] can be provided to an existing [SharedElementTransitionScope] using
+ * [ProvideAnimatedTransitionScope]. Within Circuit a NavigableCircuitContent using an
+ * AnimatedNavDecoration will provide a [Navigation] scope for sharing elements across Screen
+ * transitions. If Circuit overlays are used an [Overlay] scope will be provided for overlay
+ * transitions.
+ */
+@OptIn(ExperimentalSharedTransitionApi::class)
+public interface SharedElementTransitionScope : SharedTransitionScope {
+
+  @DelicateCircuitSharedElementsApi public val hasLayoutCoordinates: Boolean
+
+  /** Get the set of available [AnimatedScope]s for this [SharedElementTransitionScope]. */
+  public fun availableScopes(): Set<AnimatedScope>
+
+  /** Get the [AnimatedVisibilityScope] for the given [key], if it exists. */
+  public fun getAnimatedScope(key: AnimatedScope): AnimatedVisibilityScope?
+
+  /** Require the [AnimatedVisibilityScope] for the given [key]. Throwing if it does not. */
+  public fun requireAnimatedScope(key: AnimatedScope): AnimatedVisibilityScope {
+    return requireNotNull(getAnimatedScope(key)) { "No AnimatedVisibilityScope found for $key" }
```

Comment:
nit: should actually be `checkNotNull`. require will throw an illegalargumentexception, which I suspect isn't quite semantically what you're aiming for here